Cluster scaling (#2370)
giantswarm/docs: src/content/tutorials/fleet-management/cluster-management/cluster-autoscaler/index.md

---
linkTitle: Cluster autoscaler
title: Advanced cluster autoscaler configuration
description: Here we describe how you can customize the configuration of the managed cluster autoscaler service in your workload clusters.
weight: 90
menu:
  principal:
    parent: tutorials-fleet-management-clusters
    identifier: tutorials-fleet-management-clusters-cluster-autoscaler
user_questions:
  - Where can I find the ConfigMap to configure cluster-autoscaler?
  - What cluster-autoscaler options can I configure?
last_review_date: 2024-12-13
owner:
  - https://github.com/orgs/giantswarm/teams/team-phoenix
---

> In Giant Swarm platform, your workload clusters come with default autoscaling functionality. Today it's supported by {{/*% autoscaling_supported_versions */%}}, but our goal is bringing this feature to all supported providers.

Author: ```suggestion
In Giant Swarm platform, your workload clusters come with default autoscaling functionality. Today, it's supported by {{/*% autoscaling_supported_versions*/%}}, but our goal is to bring this feature to all supported providers.
```

---
linkTitle: Cluster autoscaler
title: Advanced cluster autoscaler configuration
description: Here we describe how you can customize the configuration of the managed cluster autoscaler service in your workload clusters.
weight: 90
menu:
  principal:
    parent: tutorials-fleet-management-clusters

>     identifier: tutorials-fleet-management-clusters-cluster-autoscaler

Reviewer: Please update from main. These menu identifiers need adjustment to latest changed.

Reviewer: Ignore me, I was looking at the wrong section.

user_questions:
  - Where can I find the ConfigMap to configure cluster-autoscaler?
  - What cluster-autoscaler options can I configure?
last_review_date: 2024-11-29
owner:
  - https://github.com/orgs/giantswarm/teams/team-phoenix
---

Your Giant Swarm installation comes with a default configuration for the [cluster-autoscaler addon](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler). You need access to the platform API to configure the cluster-autoscaler. [Learn how to access the platform API]({{< relref "/getting-started/access-to-platform-api" >}}).

You can override these defaults in a `ConfigMap` named `cluster-autoscaler-user-values`.

## Where is the user values ConfigMap

The following examples assume the cluster you are trying to configure has an id of `123ab`.

> You will find the `cluster-autoscaler-user-values` `ConfigMap` on the organization namespace of your cluster:

Reviewer: ```suggestion
You will find the ConfigMap named `cluster-autoscaler-user-values` in the organization namespace of your cluster:
```

```text
$ kubectl -n org-company get cm 123ab-cluster-autoscaler-user-values
NAME                                       DATA      AGE
123ab-cluster-autoscaler-user-values         0         11m
```

## How to set configuration options using the user values ConfigMap

> On the platform API, create or edit a `ConfigMap` named `<CLUSTER_ID>-cluster-autoscaler-user-values`

Reviewer: ```suggestion
On the platform API, create or edit a ConfigMap named `<CLUSTER_ID>-cluster-autoscaler-user-values`
```

in the workload cluster namespace:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app: cluster-autoscaler
  name: abc12-cluster-autoscaler-user-values
  namespace: myorg
data:
  values: |
    configmap:
      scaleDownUtilizationThreshold: 0.30
```

> ## Configuration Reference

Reviewer: ```suggestion
## Configuration reference
```

> The following sections explain some of the configuration options and what their defaults are. They show only the 'data' field of the ConfigMap for brevity.

Reviewer: ```suggestion
The following sections explain some of the configuration options and what their defaults are. They show only the `data` field of the ConfigMap for brevity.
```

The most recent source of truth for these values can be found in the [values.yaml](https://github.com/giantswarm/cluster-autoscaler-app/blob/v1.30.3-gs1/helm/cluster-autoscaler-app/values.yaml) file of the `cluster-autoscaler-app`.

### Scale down utilization threshold

> The `scaleDownUtilizationThreshold` defines the proportion between requested resources and capacity, which under the value `cluster-autoscaler` will trigger the scaling down action.

Reviewer: ```suggestion
The `scaleDownUtilizationThreshold` defines the proportion between requested resources and capacity. Once utilization drops below this value, cluster-autoscaler will consider a node as removable.
```

Our default value is 65%, which means in order to scale down, one of the nodes has to have less utilization (CPU/memory) than this threshold.

```yaml
# 9.0.1 and greater
data:
  values: |
    configmap:
      scaleDownUtilizationThreshold: 0.65

# 9.0.0 and below
data:
  scaleDownUtilizationThreshold: 0.65
```

> ### Scan Interval

Reviewer: Please use sentence case

```suggestion
### Scan interval
```

> Define what interval is used to review the state for taking a decision to scale up/down. Our default value is 10 seconds.

Reviewer: ```suggestion
Defines what interval is used to review the state for taking a decision to scale up/down. Our default value is 10 seconds.
```

Is this true? 10 seconds appears incredibly short to me.

```yaml
data:
  values: |
    configmap:
      scanInterval: "100s"
```

### Skip system pods

> By default, the `cluster-autoscaler` will never delete nodes which run pods of the `kube-system` namespace (except `daemonset` pods). It can be modified by setting following property to `"false"`.

Reviewer: ```suggestion
By default, the cluster autoscaler will never delete nodes which run pods of the `kube-system` namespace (except `daemonset` pods). This rule can be deactivated by setting the following property to false.
```

```yaml
data:
  values: |
    configmap:
      skipNodesWithSystemPods: "false"
```

### Skip pods with local storage

> The `cluster-autoscaler` configuration by default deletes nodes with pods using local storage (`hostPath` or `emptyDir`). In case you want to disable this action, you need to set the following property to `"true"`.

Reviewer: ```suggestion
The cluster autoscaler by default deletes nodes with pods using local storage (`hostPath` or `emptyDir`). In case you want to protect these nodes from removal, you can to set the following property to true.
```

```yaml
data:
  values: |
    configmap:
      skipNodesWithLocalStorage: "true"
```

### Balance similar node groups

The `cluster-autoscaler` configuration by default doesn't differentiate between node groups when scaling. In case you want to enable this action, you need to set the following property to `"true"`.

```yaml
data:
  values: |
    configmap:
      balanceSimilarNodeGroups: "true"
```

> Read [the `Kubernetes` autoscaler FAQ](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md) to learn more about the `cluster-autoscaler` and its configuration options.

Reviewer: No need for literals here

```suggestion
Read [the Kubernetes autoscaler FAQ](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md) to learn more about the cluster autoscaler and its configuration options.
```

    identifier: tutorials-fleet-management-clusters-cluster-autoscaler
user_questions:
  - Where can I find the ConfigMap to configure cluster-autoscaler?
  - What cluster-autoscaler options can I configure?
last_review_date: 2024-11-29
owner:
  - https://github.com/orgs/giantswarm/teams/team-phoenix
---

Your Giant Swarm installation comes with a default configuration for the [cluster-autoscaler addon](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler). You need access to the platform API to configure the cluster-autoscaler. [Learn how to access the platform API]({{< relref "/getting-started/access-to-platform-api" >}}).

You can override these defaults in a `ConfigMap` named `cluster-autoscaler-user-values`.

## Where is the user values ConfigMap

The following examples assume the cluster you are trying to configure has an id of `123ab`.

You will find the `cluster-autoscaler-user-values` `ConfigMap` on the organization namespace of your cluster:

```text
$ kubectl -n org-company get cm 123ab-cluster-autoscaler-user-values
NAME                                       DATA      AGE
123ab-cluster-autoscaler-user-values         0         11m
```

> 

Author: ```suggestion
```

    identifier: tutorials-fleet-management-clusters-cluster-autoscaler
user_questions:
  - Where can I find the ConfigMap to configure cluster-autoscaler?
  - What cluster-autoscaler options can I configure?
last_review_date: 2024-11-29
owner:
  - https://github.com/orgs/giantswarm/teams/team-phoenix
---

Your Giant Swarm installation comes with a default configuration for the [cluster-autoscaler addon](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler). You need access to the platform API to configure the cluster-autoscaler. [Learn how to access the platform API]({{< relref "/getting-started/access-to-platform-api" >}}).

You can override these defaults in a `ConfigMap` named `cluster-autoscaler-user-values`.

## Where is the user values ConfigMap

The following examples assume the cluster you are trying to configure has an id of `123ab`.

You will find the `cluster-autoscaler-user-values` `ConfigMap` on the organization namespace of your cluster:

```text
$ kubectl -n org-company get cm 123ab-cluster-autoscaler-user-values
NAME                                       DATA      AGE
123ab-cluster-autoscaler-user-values         0         11m
```

> On cluster creation the user values `ConfigMap` is empty (or might not exist yet) and the following defaults will be applied to the final `cluster-autoscaler` deployment. To customize any of the configuration options, you just need to add the respective lines in the data field of the user `ConfigMap`.

Author: ```suggestion
```

---
linkTitle: Cluster autoscaler
title: Advanced cluster autoscaler configuration
description: Here we describe how you can customize the configuration of the managed cluster autoscaler service in your workload clusters.
weight: 90
menu:
  principal:
    parent: tutorials-fleet-management-clusters
    identifier: tutorials-fleet-management-clusters-cluster-autoscaler
user_questions:
  - Where can I find the ConfigMap to configure cluster-autoscaler?
  - What cluster-autoscaler options can I configure?
last_review_date: 2024-11-29
owner:
  - https://github.com/orgs/giantswarm/teams/team-phoenix
---

> Your Giant Swarm installation comes with a default configuration for the [cluster-autoscaler addon](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler). You need access to the platform API to configure the cluster-autoscaler. [Learn how to access the platform API]({{< relref "/getting-started/access-to-platform-api" >}}).

Reviewer 2: Currently, the `cluster-autoscaler` is running on the WC, so there is one per WC, instead of one `cluster-autoscaler` per installation. Also, not all providers currently support `cluster-autoscaler`, i.e CAPZ does not support `cluster-autoscaler`.

> You can override these defaults in a `ConfigMap` named `cluster-autoscaler-user-values`.

Reviewer 2: The name of the configmap is prefixed with the name of the WC, like `golem-cluster-autoscaler-user-values`.

## Where is the user values ConfigMap

The following examples assume the cluster you are trying to configure has an id of `123ab`.

You will find the `cluster-autoscaler-user-values` `ConfigMap` on the organization namespace of your cluster:

```text

> $ export ORG=yourorg
> $ export CLUSTER=123ab
> $ kubectl -n $ORG get cm $CLUSTER-cluster-autoscaler-user-values
> NAME                                       DATA      AGE
> 123ab-cluster-autoscaler-user-values         0         11m

Reviewer 2: nitpick: I find this confusing TBH. I'd simplify it to be just

```suggestion
$ kubectl -n org-company get cm 123ab-cluster-autoscaler-user-values
NAME                                       DATA      AGE
123ab-cluster-autoscaler-user-values         0         11m
```
